sys/net/application_layer/gcoap/gcoap.c

```diff
@@ -123,150 +121,130 @@ static void *_event_loop(void *arg)
+            _find_req_memo(&memo, &pdu, &remote);
+            if (memo) {
+                switch (coap_get_type(&pdu)) {
+                case COAP_TYPE_NON:
+                case COAP_TYPE_ACK:
+                    if (&memo->resp_evt_tmout.queue) {
+                        event_timeout_clear(&memo->resp_evt_tmout);
+                    }
+                    memo->state = GCOAP_MEMO_RESP;
+                    if (memo->resp_handler) {
+                        memo->resp_handler(memo, &pdu, &remote);
+                    }
+
+                    if (memo->send_limit >= 0) {        /* if confirmable */
+                        *memo->msg.data.pdu_buf = 0;    /* clear resend PDU buffer */
+                    }
+                    memo->state = GCOAP_MEMO_UNUSED;
+                    break;
+                case COAP_TYPE_CON:
+                    DEBUG("gcoap: separate CON response not handled yet\n");
+                    break;
+                default:
+                    DEBUG("gcoap: illegal response type: %u\n", coap_get_type(&pdu));
+                    break;
                 }
-                memo->state = GCOAP_MEMO_UNUSED;
-                break;
-            case COAP_TYPE_CON:
-                DEBUG("gcoap: separate CON response not handled yet\n");
-                break;
-            default:
-                DEBUG("gcoap: illegal response type: %u\n", coap_get_type(&pdu));
-                break;
             }
+            else {
+                DEBUG("gcoap: msg not found for ID: %u\n", coap_get_id(&pdu));
+            }
+            break;
+        default:
+            DEBUG("gcoap: illegal code class: %u\n", coap_get_code_class(&pdu));
         }
-        else {
-            DEBUG("gcoap: msg not found for ID: %u\n", coap_get_id(&pdu));
+    }
+}
+
+/* Handles response timeout for a request; resend confirmable if needed. */
+static void _on_resp_timeout(void *arg) {
+    gcoap_request_memo_t *memo = (gcoap_request_memo_t *)arg;
+
+    /* no retries remaining */
+    if ((memo->send_limit == GCOAP_SEND_LIMIT_NON) || (memo->send_limit == 0)) {
+        _expire_request(memo);
+    }
+    /* reduce retries remaining, double timeout and resend */
+    else {
+        memo->send_limit--;
+#ifdef CONFIG_GCOAP_NO_RETRANS_BACKOFF
+        unsigned i        = 0;
+#else
+        unsigned i        = COAP_MAX_RETRANSMIT - memo->send_limit;
+#endif
+        uint32_t timeout  = ((uint32_t)COAP_ACK_TIMEOUT << i) * US_PER_SEC;
+#if COAP_RANDOM_FACTOR_1000 > 1000
+        uint32_t end = ((uint32_t)TIMEOUT_RANGE_END << i) * US_PER_SEC;
+        timeout = random_uint32_range(timeout, end);
+#endif
+        event_timeout_set(&memo->resp_evt_tmout, timeout);
+
+        ssize_t bytes = sock_udp_send(&_sock, memo->msg.data.pdu_buf,
+                                      memo->msg.data.pdu_len, &memo->remote_ep);
+        if (bytes <= 0) {
+            DEBUG("gcoap: sock resend failed: %d\n", (int)bytes);
+            _expire_request(memo);
```

Comment:
Note to self: This was adapted from `_event_loop`'s `GCOAP_MSG_TYPE_TIMEOUT` case.